### PR TITLE
Daemon/PolicyAdd lock policyRepo to avoid fqdn races.

### PR DESCRIPTION
**Pull request by eloycoto**
This is an small backport to fix the FQDN race found in the issue #7105
and largely discussed in PR #7220. Due in master a new behaviour will be
used that cannot be backported to 1.4, fix the issue in 1.4 with a
specific commit.

This commit will lock PolicyRepo in the start, so rules are never going
to be overwritted by Fqdn GeneratedRule callback.

PR: https://github.com/cilium/cilium/pull/7220

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7390)

### CODE REVIEW

**Comment by eloycoto** [comment]
test-me-please

File: daemon/policy.go [hunk 1]
```diff
@@ -215,6 +218,7 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (newRev uint
 			err2 := fmt.Errorf("Unable to recompile base programs: %s", err)
 			log.WithError(err2).WithField("prefixes", prefixes).Warn(
 				"Failed to recompile base programs due to prefix length count change")
+			d.policy.Mutex.Unlock()
```
Comment (aanm) [inline daemon/policy.go]: Do we really need to have policy lock while compiling a bpf program?

Comment (ianvernon) [reply daemon/policy.go]: Because FQDN policy generates CIDR rules (which have prefix lengths), then yes, we will have to hold the lock. However, I believe that this would only occur upon the first CIDR rule which contains a /32 which didn't before (@joestringer can you confirm?). We could, in `v1.4` always keep a `/32` prefix length around to avoid this issue. FWIW, this need to compile base programs will only occur for kernels >= 4.9 <= 4.11.

**Comment by eloycoto** [comment]
test-me-please

File: daemon/policy.go (hunk 1 above)
Comment (joestringer) [reply daemon/policy.go]: We already keep a `/32` around all the time anyway since we're guaranteed to need to refer to `/32s`:

https://github.com/cilium/cilium/blob/dc8a343a260594a1a95ff5ede63cd13f8c97f81b/daemon/daemon.go#L842-L855

However this function also handles shorter CIDRs, which could end up needing recompilation, which I think would be required to go into this particular conditional block.

Comment (joestringer) [reply daemon/policy.go]: In other words, yes, we need it, but also this should never occur for FQDN, only when a CIDR with a new unique prefix length is added.

Comment (joestringer) [reply daemon/policy.go]: But coming back to @aanm's question, I'm not sure I fully understand why recompilation of base programs needs to be done while holding the policy lock. This affects ipcache lookups for the base programs like the ones attached to the `cilium_host` device, not endpoint devices (which would be handled through individual endpoint BPF program compilation).

Comment (joestringer) [reply daemon/policy.go]: Ideally we don't compile the base programs and then later on find out that we're not going to apply the policy and need to recompile the base programs again to back out the policy.. incidentally we actually skip this part when rolling back the changes in the case that `ipcache.AllocateCIDRs(...)` fails. Not strictly right but also it will only have a pretty small performance impact.

TL;DR I noticed a case we don't properly revert out but it doesn't really matter.

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-missed-k8s

File: daemon/policy.go (hunk 1 above)
Comment (ianvernon) [reply daemon/policy.go]: Great, having '/32' always around solves the issue , thanks for confirming Joe!

**Review by ianvernon — APPROVED** [review]

**Comment by ianvernon** [comment]
test-docs-please

**Comment by ianvernon** [comment]
@eloycoto is this ready for merging? It is still `wip`, but all tests passed.

**Comment by eloycoto** [comment]
yep, I think that should be ok.

**Review by tgraf — APPROVED** [review]
